sections/03_testing_code.qmd

```diff
@@ -1,3 +1,13 @@
--   Unit tests
--   Integration tests
--   Test-driven development
+When we are conducting experiments we often conduct rigorous testing to ensure that results are reliable and accurate - why should that be any different when you are developing your code? By testing your code you can be sure that it is working as intended, as well as identify any limitations of the code. While the idea of developing tests may feel a bit excessive when starting out with coding, it is valuable to be aware of the principles as it still provides a conceptual basis from which you can develop code that meets the expectations associated with conducting 'good' science.
+
+### Unit tests
+
+Unit testing focuses on testing *individual functions* of your codebase to ensure that a function is doing what it should be doing and meets the specified requirements in a formalised and automated manner. At a high level the aim of unit tests is to make sure that the underlying maths of your function is correct. This can be done by inputting a value into the function for which you know what the output (answer) is and testing if the answer that the function gives you is the same.
+
+### Integration tests
+
+Integration tests are more about ensuring that the parts fit into the whole. So going back to the data analysis example - we want to make sure that the output from our data cleaning function is able to seamlessly act as the input for our data analysis function. Alternatively you might want to run integration tests when you are introducing new features (functions) to your project and need to make sure that these do not break or alter the behaviour of your existing workflow.
+
+### Test-driven development
```

Comment:
I'm also not 100% sold on this section (falls under the docker TDD umbrella) but I think we can follow the trend of keeping it here and seeing what others think